src/test/java/com/ericsson/eiffelcommons/jenkins/JenkinsManagerTest.java

```diff
@@ -0,0 +1,614 @@
+        jenkins.getJenkinsBuildStatusData(JOB_NAME, BUILD_NUMBER);
+    }
+
+    @Test
+    public void getJenkinsBuildStatusDataOnlyJobName() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpStatusEndpointFallback();
+        JSONObject json = jenkins.getJenkinsBuildStatusData(JOB_NAME);
+        assertEquals(BODY, json.toString());
+    }
+
+    @Test
+    public void pluginExists() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginEndpoint();
+        boolean success = jenkins.pluginExists(PLUGIN_NAME);
+        assertTrue(success);
+    }
+
+    @Test
+    public void pluginExistsNotFound() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginEndpoint();
+        boolean success = jenkins.pluginExists(PLUGIN_NOT_FOUND);
+        assertFalse(success);
+    }
+
+    @Test(expected = Exception.class)
+    public void pluginExistsFailRequest() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginEndpointFail();
+        jenkins.pluginExists(PLUGIN_NAME);
+    }
+
+    @Test
+    public void installPlugin() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginInstallEndpoint();
+        boolean success = jenkins.installPlugin(PLUGIN_NAME, PLUGIN_VERSION);
+        assertTrue(success);
+    }
+
+    @Test(expected = Exception.class)
+    public void installPluginNameNull() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginInstallEndpoint();
+        jenkins.installPlugin(null, PLUGIN_VERSION);
+    }
+
+    @Test(expected = Exception.class)
+    public void installPluginNameEmpty() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginInstallEndpoint();
+        jenkins.installPlugin("", PLUGIN_VERSION);
+    }
+
+    @Test(expected = Exception.class)
+    public void installPluginVersionNull() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginInstallEndpoint();
+        jenkins.installPlugin(PLUGIN_NAME, null);
+    }
+
+    @Test(expected = Exception.class)
+    public void installPluginVersionEmpty() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginInstallEndpoint();
+        jenkins.installPlugin(PLUGIN_NAME, "");
+    }
+
+    @Test(expected = Exception.class)
+    public void installPluginFailRequest() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpPluginInstallEndpointFail();
+        jenkins.installPlugin(PLUGIN_NAME, PLUGIN_VERSION);
+    }
+
+    @Test
+    public void restartJenkinsServerNeverDown() throws Exception {
+        JenkinsManager jenkins = setUpJenkinsManager();
+        setUpRestartAndAvailableEndpoint();
+        boolean success = jenkins.restartJenkins();
+        assertFalse(success);
```

Comment:
why do we return false here? Or am I just confused because of the method name maybe?

Reply:
This is an odd one and I was also confused but it works like this, if the verify method does not detect that the jenkins was down on the first check it will determine that jenkins never went down and thus restart technically "failed". It will give true if it first detects that jenkins is unavailable and then later is ok.

Reply:
Yes, JenkinsManager will ensure that a restart was successfull by checking that it did indeed become unavailable for a time and then available. If it never became unavailable it assumes it never restarted. I think we do this in System tests, but not sure we validate it. A restart must be done after installing a plugin.

Reply:
Can we place this knowledge in a comment so the next reader is not confused?

Reply:
I added a comment.

Reply:
Good